### PR TITLE
Fix to close connection upon exception during waiting for greetings

### PR DESCRIPTION
**Pull request by cifer828**
## Description

Fix a couple of exception cases that we miss to close the connection.

## Motivation and Context

Channels should be closed in all cases to prevent memory leaking.

## How Has This Been Tested?

Add new test cases to cover methods for closing channel.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Major release (change is NOT backward compatible with prior release)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/test/java/com/yahoo/smtpnio/async/client/SmtpAsyncClientTest.java [hunk 1]
```diff
@@ -701,6 +711,93 @@ public void testCreateSessionNoLocalAddressSNIEmptySuccessful() throws SSLExcept
...
+        final ArgumentCaptor<ChannelHandler> handlerCaptorFirst = ArgumentCaptor.forClass(ChannelHandler.class);
+
+        Mockito.verify(nettyPipeline, Mockito.times(0)).addFirst(Mockito.anyString(), handlerCaptorFirst.capture());
+        Assert.assertEquals(handlerCaptorFirst.getAllValues().size(), 0, "number of handlers mismatched.");
+        final ArgumentCaptor<ChannelHandler> handlerCaptorLast = ArgumentCaptor.forClass(ChannelHandler.class);
+        Mockito.verify(nettyPipeline, Mockito.times(0)).addLast(Mockito.anyString(), handlerCaptorLast.capture());
+        Assert.assertEquals(handlerCaptorLast.getAllValues().size(), 0, "Unexpected count of ChannelHandler added.");
+
+        // verify logging messages
+        // verify logging messages
```
Comment (fansu) [inline core/src/test/java/com/yahoo/smtpnio/async/client/SmtpAsyncClientTest.java:774]: remove this line

File: core/src/test/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandlerTest.java [hunk 2]
```diff
@@ -200,10 +253,15 @@ public void testUserEventTriggeredIdleStateEventReadIdle() throws IllegalArgumen
...
         Assert.assertEquals(aEx.getFailureType(), FailureType.CONNECTION_FAILED_EXCEED_IDLE_MAX, "Failure type mismatched");
 
+        Mockito.verify(ctx, Mockito.times(1)).close();
+        Mockito.verify(channel, Mockito.times(1)).isActive();
+
         // call channelInactive, should not encounter npe
         handler.channelInactive(ctx);
     }
 
+
```
Comment (fansu) [inline core/src/test/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandlerTest.java:263]: remove

File: core/src/test/java/com/yahoo/smtpnio/async/client/SmtpAsyncClientTest.java [hunk 3]
```diff
@@ -701,6 +711,93 @@ public void testCreateSessionNoLocalAddressSNIEmptySuccessful() throws SSLExcept
...
+            Assert.assertEquals(exception.getClass(), SmtpAsyncClientException.class, "exception type mismatch." + ex);
+            actual = (SmtpAsyncClientException) exception;
+        }
+
+        Assert.assertNotNull(actual.getCause(), "Cause should not be null");
+        Assert.assertEquals(actual.getCause().getClass(), ConnectTimeoutException.class, "Cause should be connection timeout exception");
+        Assert.assertSame(actual.getCause(), nettyConnectFuture.cause(), "Cause should be same object");
+        Assert.assertEquals(actual.getFailureType(), FailureType.WRITE_TO_SERVER_FAILED, "Exception type should be CONNECTION_TIMEOUT_EXCEPTION");
+
+
```
Comment (fansu) [inline core/src/test/java/com/yahoo/smtpnio/async/client/SmtpAsyncClientTest.java:798]: remove

**Comment by fansu** [comment]
@jui8wang

File: core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java [hunk 4]
```diff
@@ -173,6 +175,17 @@ public long applyAsLong(final long counter) { // increment by 1 unless it overfl
...
     }
 
+    /**
+     * Closes channel.
+     *
+     * @param channel the channel
+     */
+    private void closeChannel(@Nullable final Channel channel) {
+        if (channel != null && channel.isActive()) {
+            channel.close();
```
Comment (jui8wang) [inline core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java:185]: do u have a test case that verifies that close is called?

File: core/src/main/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandler.java [hunk 5]
```diff
@@ -82,6 +94,7 @@ public void decode(@Nonnull final ChannelHandlerContext ctx, @Nonnull final Smtp
             logger.error("[{},{}] Server response was not successful:{}", sessionId, sessionCtx, serverResponse.toString());
             sessionCreatedFuture.done(new SmtpAsyncClientException(FailureType.CONNECTION_FAILED_INVALID_GREETING_CODE, sessionId, sessionCtx,
                     serverResponse.toString()));
+            close(ctx); // closing the channel if we r not getting a ok greeting
```
Comment (jui8wang) [inline core/src/main/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandler.java:97]: same here, do u have a test case that verifies it closes the channel

File: core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java [hunk 6]
```diff
@@ -133,6 +134,7 @@ public void operationComplete(final io.netty.util.concurrent.Future<? super Void
                             future.cause());
                     sessionFuture.done(ex);
                     logger.error(CONNECT_RESULT_REC, "N/A", sessionCtx, "failure", host, port, enableSsl, sniNames, ex);
+                    closeChannel(nettyConnectFuture.channel());
```
Comment (jui8wang) [inline core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java:137]: Could u go through ImapAsyncSessionImpl,  and see if i have added close channel calls upon error?  If true, and smtp does not have it, please add.  Please just double check.

File: core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java (hunk 4 above)
Comment (cifer828) [reply core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java:185]: yes, testCreateSessionFailedWithSSL, testCreateSessionFailedWithNoSSL and testCreateSessionConnectionFailedChannelIsNull verifies close() and covers the branches.  100% coverage for the methods added in this PR.

File: core/src/main/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandler.java (hunk 5 above)
Comment (cifer828) [reply core/src/main/java/com/yahoo/smtpnio/async/netty/SmtpClientConnectHandler.java:97]: yes, testExceptionCaughtChannelWasClosedAlready, testExceptionCaught, testDecodeConnectFailed and testUserEventTriggeredIdleStateEventReadIdle verify close() in different situation.

File: core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java (hunk 6 above)
Comment (cifer828) [reply core/src/main/java/com/yahoo/smtpnio/async/client/SmtpAsyncClient.java:137]: checked. SmtpAsyncSessionImpl uses the same close strategy as ImapAsyncSessionImpl. where imap throws an exception and close the channel, smtp does the same thing.

**Comment by fansu** [comment]
@nituyahoo

**Pull request merged**